### ISSUE TITLE
Fierce Pearl Falcon - Incorrect Implementation of Upgradability Pattern

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-09-Gitcoin-judging#787]
Fierce Pearl Falcon

medium

# Incorrect Implementation of Upgradability Pattern
## Summary

By inheriting from non-upgradeable OpenZeppelin contracts, the subsequent versions of the contract can become corrupt.

## Vulnerability Detail

The Allo and Registry contracts employ an upgradability pattern but make the mistake of inheriting from non-upgradeable OpenZeppelin contracts.

        import "openzeppelin-contracts/contracts/access/AccessControl.sol";

https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/core/Allo.sol#L8
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/core/Registry.sol#L5

The contracts should inherit from `AccessControlUpgradeable` rather than `AccessControl`.

OpenZeppelin's non-upgradeable contracts are not designed to support future upgrades. Specifically, they lack the `__gap` variable that helps reserve storage slots for future versions to prevent storage layout conflicts during the inheritance chain.

        /**
        * @dev This empty reserved space is put in place to allow future versions to add new
        * variables without shifting down storage in the inheritance chain.
        * See https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
        */
        uint256[50] private __gap;

https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/bc95521e34dcd49792065e264a7ad2b5a86f0091/contracts/access/AccessControlCrossChainUpgradeable.sol#L52-L57

## Impact

The contract storage can be corrupted if a new variable is added to the parent contract in future iterations.

## Code Snippet

https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/core/Allo.sol#L8
https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/core/Registry.sol#L5

## Tool used

Manual Review

## Recommendation

The contract should inherent from `AccessControlUpgradeable` instead of `AccessControl`.

```diff
- import "openzeppelin-contracts/contracts/access/AccessControl.sol";
+ import "openzeppelin-contracts-upgradeable/contracts/access/AccessControlUpgradeable.sol";
```

**Comment by thangtranth** [comment]
Escalate

Contract storage will be corrupted and it breaks the upgradable ability of the contracts. This is medium issue in my opinion.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> Contract storage will be corrupted and it breaks the upgradable ability of the contracts. This is medium issue in my opinion. 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by 0x00ffDa** [comment]
FYI, duplicate of https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/852

**Comment by neeksec** [comment]
Low/info because of Sherlock's rule,

>Use of Storage gaps: Simple contracts with one of the parent contract not implementing storage gaps are considered low/informational. 
Exception: However, if the protocol design has a highly complex and branched set of contract inheritance with storage gaps inconsistently applied throughout and the submission clearly describes the necessity of storage gaps it can be considered a valid medium. [Example](https://github.com/sherlock-audit/2022-09-notional-judging/issues/64)

This is a duplicate of #852. You can also check https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/852#issuecomment-1769119628 for my reasons why it's low.

**Comment by thangtranth** [comment]
#852 has not mentioned anything about the necessity of the storage gap. 

For example in the Registry contract: https://github.com/sherlock-audit/2023-09-Gitcoin/blob/main/allo-v2/contracts/core/Registry.sol#L5

The storage layout is:

// From Initializable
uint8 private _initialized;
bool private _initializing;
// From AccessControl
mapping(bytes32 => RoleData) private _roles;
// From Registry
mapping(address => bytes32) public anchorToProfileId;
mapping(bytes32 => Profile) public profilesById;
mapping(bytes32 => address) public profileIdToPendingOwner;

Therefore the AccessControl can never be upgraded to a new version with new storage variables.

---

If using correct AccessControlUpgradeable contract, the layout will support all the imaginable upgrade in the future in the AccessControl base contract:

// From Initializable
uint8 private _initialized;
bool private _initializing;
// From AccessControl
mapping(bytes32 => RoleData) private _roles;
**<49 bytes32 slots for new variables>**
// From Registry
mapping(address => bytes32) public anchorToProfileId;
mapping(bytes32 => Profile) public profilesById;
mapping(bytes32 => address) public profileIdToPendingOwner;

So the issue prevent upgradability in the base contract.

**Comment by Evert0x** [comment]
> Therefore the AccessControl can never be upgraded to a new version with new storage variables.

This is just explaining the behavior of using `AccessControl` instead of `AccessControlUpgradeable`, right?

Planning to reject escalation as it seems informational to me.

@thangtranth

**Comment by thangtranth** [comment]
@Evert0x : Yeah, I think it is ok for it to be informational since no fund is at risk when the contract is deployed. The important thing is that the issue get fixed by the sponsor. 

Thanks.

**Comment by Evert0x** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [thangtranth](https://github.com/sherlock-audit/2023-09-Gitcoin-judging/issues/787/#issuecomment-1764013085): rejected
